src/routes/data.rs

```diff
@@ -217,3 +218,20 @@ pub async fn retrieve_card(
 
     Ok(Json(card.try_into()?))
 }
+
+/// `/cards/fingerprint` handling the creation and retrieval of card fingerprint
+pub async fn get_or_insert_fingerprint(
+    extract::State(state): extract::State<AppState>,
+    Json(request): Json<types::FingerprintRequest>,
+) -> Result<Json<types::FingerprintResponse>, ContainerError<error::ApiError>> {
+    crate::validations::luhn_on_string(request.card.card_number.inner().peek())
+        .then_some(())
+        .ok_or(error::ApiError::ValidationError("card number invalid"))?;
```

Comment:
Let's keep is declarative, instead of keeping this here, put it inside the `Validate` trait